tailscale/client.go

```diff
@@ -45,6 +53,45 @@ type (
 const baseURL = "https://api.tailscale.com"
 const contentType = "application/json"
 
+// NewClientFromOAuth returns a new instance of the Client type that will perform operations against a chosen tailnet using
+// OAuth Credentials. The Client will generate a new API key against the Tailscale OAuth token endpoint. The API key will be
+// renewed based on the `expires_in` property returned from the Tailscale OAuth token endpoint.
+// Additional options can be provided, see ClientOption for more details.
+func NewClientFromOAuth(ctx context.Context, oauthClientID, oauthClientSecret string, oauthScopes []string, tailnet string, options ...ClientOption) (*Client, error) {
+	u, err := url.Parse(baseURL)
+	if err != nil {
+		return nil, err
+	}
+
+	var oauthConfig = &clientcredentials.Config{
+		ClientID:     oauthClientID,
+		ClientSecret: oauthClientSecret,
+		TokenURL:     fmt.Sprintf("%s/api/v2/oauth/token", baseURL),
+		Scopes:       oauthScopes,
+	}
+
+	c := &Client{
+		oauthConfig: oauthConfig,
+		http:        httpClientWithDefaults(),
+		baseURL:     u,
+		tailnet:     tailnet,
+	}
+
+	// return an error message at client initialization if oauth credentials fail instead of waiting for first request
+	_, err = c.apiKeyFromOAuth(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error getting API key from OAuth token endpoint: %w", err)
+	}
```

Comment:
I think we'd want to do this after applying `options`, just before returning from this function.